develop/dev-guide-sample-application-python.md

```diff
@@ -0,0 +1,1103 @@
+        return False
+
+    buy_coins = Player.select(Player.coins).where(Player.id == buy_id).get().coins
+    if buy_coins < price:
+        print(f'buy player {buy_id} coins not enough')
+        return False
+
+    return True
+
+
+def trade(sell_id: str, buy_id: str, amount: int, price: int) -> None:
+    with db.atomic() as txn:
+        try:
+            if trade_check(sell_id, buy_id, amount, price) is False:
+                txn.rollback()
+                return
+
+            # deduct the goods of seller, and raise his/her the coins
+            Player.update(goods=Player.goods - amount, coins=Player.coins + price).where(Player.id == sell_id).execute()
+            # deduct the coins of buyer, and raise his/her the goods
+            Player.update(goods=Player.goods + amount, coins=Player.coins - price).where(Player.id == buy_id).execute()
+
+        except Exception as err:
+            txn.rollback()
+            print(f'something went wrong: {err}')
+        else:
+            txn.commit()
+            print("trade success")
+
+
+def trade_example() -> None:
+    # create two players
+    # player 1: id is "1", has only 100 coins.
+    # player 2: id is "2", has 114514 coins, and 20 goods.
+    Player.create(id="1", coins=100, goods=0)
+    Player.create(id="2", coins=114514, goods=20)
+
+    # player 1 wants to buy 10 goods from player 2.
+    # it will cost 500 coins, but player 1 cannot afford it.
+    # so this trade will fail, and nobody will lose their coins or goods
+    trade(sell_id="2", buy_id="1", amount=10, price=500)
+
+    # then player 1 has to reduce the incoming quantity to 2.
+    # this trade will be successful
+    trade(sell_id="2", buy_id="1", amount=2, price=100)
+
+    # let's take a look for player 1 and player 2 currently
+    after_trade_players = Player.select().where(Player.id.in_(["1", "2"]))
+    for player in after_trade_players:
+        print(f'id:{player.id}, coins:{player.coins}, goods:{player.goods}')
+
+
+db.connect()
+
+# recreate the player table
+db.drop_tables([Player])
+db.create_tables([Player])
+
+simple_example()
+trade_example()
+```
+
+Compared with using drivers directly, peewee provides an abstraction for the specific details of different databases when creating a database connection. Additionally, peewee encapsulates some operations such as session management and CRUD of basic objects, which greatly simplifies the code.
+
+The `Player` class is a mapping of a table to attributes in the program. Each attribute of `Player` corresponds to a field in the `player` table. To provide SQLAlchemy with more information, the attribute is defined as `id = Column(String(36), primary_key=True)` to indicate the field type and its additional properties. For example, `id = Column(String(36), primary_key=True)` indicates that the `id` attribute is `String` type, the corresponding field in database is `VARCHAR` type, the length is `36`, and it is a primary key.
```

Comment:
ditto

Reply:
Sorry, this `ditto` is tracked for which comment?